File: 01/TwoPassAssembler.py
# Design suitable Data structures and implement Pass-I and Pass-II of a two-pass assembler for pseudo
# machine. Implementation should consist of a few instructions from each category and few assembler 
# directives. The output of Pass-I (intermediate code file and symbol table) should be input for Pass-II.

class Assembler:
    def __init__(self):
        self.symbol_table = {}
        self.intermediate_code = []
        self.location_counter = 0

    def pass_one(self, source_code):
        for line in source_code:
            line = line.strip()
            if not line or line.startswith(';'):  # Skip empty lines and comments
                continue

            parts = line.split()
            instruction = parts[0]

            if instruction == 'START':
                self.location_counter = int(parts[1])
            elif instruction == 'END':
                continue
            elif instruction == 'WORD':
                self.intermediate_code.append((self.location_counter, 'WORD', parts[1]))
                self.location_counter += 1
            elif instruction == 'RESW':
                size = int(parts[1])
                self.intermediate_code.append((self.location_counter, 'RESW', size))
                self.location_counter += size
            else:
                # Assume it's a label or an instruction
                if ':' in instruction:
                    label = instruction[:-1]
                    self.symbol_table[label] = self.location_counter
                    instruction = parts[1]  # The next part is the actual instruction

                self.intermediate_code.append((self.location_counter, instruction, parts[1:] if len(parts) > 1 else []))
                self.location_counter += 1

    def pass_two(self):
        final_output = []
        for loc, instruction, operands in self.intermediate_code:
            if instruction == 'WORD':
                final_output.append(f"{loc}: {instruction} {operands[0]}")
            elif instruction == 'RESW':
                final_output.append(f"{loc}: {instruction} {operands[0]}")
            else:
                operand_addresses = []
                for operand in operands:
                    if operand in self.symbol_table:
                        operand_addresses.append(self.symbol_table[operand])
                    else:
                        operand_addresses.append(operand)  # For immediate values, etc.
                final_output.append(f"{loc}: {instruction} {' '.join(map(str, operand_addresses))}")

        return final_output

# Example source code
source_code = [
    "START 100",
    "LABEL1: LOAD VALUE",
    "ADD LABEL2",
    "STORE RESULT",
    "VALUE WORD 5",
    "LABEL2: WORD 10",
    "RESULT RESW 1",
    "END"
]

# Create assembler instance
assembler = Assembler()
assembler.pass_one(source_code)
final_output = assembler.pass_two()

# Display the symbol table and final output
print("Symbol Table:")
for label, address in assembler.symbol_table.items():
    print(f"{label}: {address}")

print("\nIntermediate Code / Final Output:")
for line in final_output:
    print(line)

File: 01/TwoPassAssembler.java
// Design suitable Data structures and implement Pass-I and Pass-II of a two-pass assembler for pseudo
// machine. Implementation should consist of a few instructions from each category and few assembler 
// directives. The output of Pass-I (intermediate code file and symbol table) should be input for Pass-II.

import java.util.*;

class TwoPassAssembler {
    private Map<String, Integer> symbolTable = new HashMap<>();
    private List<String> intermediateCode = new ArrayList<>();
    private int locationCounter = 0;

    public void passOne(List<String> sourceCode) {
        for (String line : sourceCode) {
            line = line.trim();
            if (line.isEmpty() || line.startsWith(";")) {
                continue; // Skip empty lines and comments
            }

            String[] parts = line.split("\\s+");
            String instruction = parts[0];

            switch (instruction) {
                case "START":
                    locationCounter = Integer.parseInt(parts[1]);
                    break;
                case "END":
                    break;
                case "WORD":
                    intermediateCode.add(locationCounter + ": WORD " + parts[1]);
                    locationCounter++;
                    break;
                case "RESW":
                    int size = Integer.parseInt(parts[1]);
                    intermediateCode.add(locationCounter + ": RESW " + size);
                    locationCounter += size;
                    break;
                default:
                    if (instruction.endsWith(":")) {
                        String label = instruction.substring(0, instruction.length() - 1);
                        symbolTable.put(label, locationCounter);
                        instruction = parts[1]; // The next part is the actual instruction
                    }
                    intermediateCode.add(locationCounter + ": " + instruction + " " + Arrays.toString(Arrays.copyOfRange(parts, 1, parts.length)));
                    locationCounter++;
            }
        }
    }

    public List<String> passTwo() {
        List<String> finalOutput = new ArrayList<>();
        for (String code : intermediateCode) {
            String[] parts = code.split(": ");
            String loc = parts[0];
            String instruction = parts[1];
            String[] operands = instruction.split(" ", 2);
            String opcode = operands[0];
            String operandPart = operands.length > 1 ? operands[1] : "";

            if (opcode.equals("WORD") || opcode.equals("RESW")) {
                finalOutput.add(code); // No change for directives
            } else {
                String[] operandArray = operandPart.split(" ");
                StringBuilder resolvedOperands = new StringBuilder();
                for (String operand : operandArray) {
                    if (symbolTable.containsKey(operand)) {
                        resolvedOperands.append(symbolTable.get(operand)).append(" ");
                    } else {
                        resolvedOperands.append(operand).append(" "); // For immediate values, etc.
                    }
                }
                finalOutput.add(loc + ": " + opcode + " " + resolvedOperands.toString().trim());
            }
        }
        return finalOutput;
    }

    public void printSymbolTable() {
        System.out.println("Symbol Table:");
        for (Map.Entry<String, Integer> entry : symbolTable.entrySet()) {
            System.out.println(entry.getKey() + ": " + entry.getValue());
        }
    }

    public static void main(String[] args) {
        List<String> sourceCode = Arrays.asList(
            "START 100",
            "LABEL1: LOAD VALUE",
            "ADD LABEL2",
            "STORE RESULT",
            "VALUE WORD 5",
            "LABEL2: WORD 10",
            "RESULT RESW 1",
            "END"
        );

        TwoPassAssembler assembler = new TwoPassAssembler();
        assembler.passOne(sourceCode);
        List<String> finalOutput = assembler.passTwo();

        assembler.printSymbolTable();
        System.out.println("\nIntermediate Code / Final Output:");
        for (String line : finalOutput) {
            System.out.println(line);
        }
    }
}
